nvflare/fuel/f3/connection.py

```diff
@@ -0,0 +1,133 @@
+conn_count = 0
+
+BytesAlike = Union[bytes, bytearray, memoryview]
+
+
+def get_connection_name():
+    global lock, conn_count
+    with lock:
+        conn_count += 1
+    return "CN%05d" % conn_count
+
+
+class ConnState(Enum):
+    IDLE = 1  # Initial state
+    CONNECTED = 2  # New connection
+    CLOSED = 3  # Connection is closed
+
+
+class FrameReceiver(ABC):
+    @abstractmethod
+    def process_frame(self, frame: BytesAlike):
+        """Frame received callback
+
+        Args:
+            frame: The frame received
+
+        Raises:
+            CommError: If any error happens while processing the frame
+        """
+        pass
+
+
+class Connection(ABC):
+    """FCI connection spec. A connection is used to transfer opaque frames"""
+
+    def __init__(self, connector: Connector):
+        self.name = get_connection_name()
+        self.state = ConnState.IDLE
+        self.frame_receiver = None
+        self.connector = connector
+
+    @abstractmethod
+    def get_conn_properties(self) -> dict:
+        """Get connection specific properties, like peer address, TLS certificate etc
+
+        Raises:
+            CommError: If any errors
+        """
+        pass
+
+    @abstractmethod
+    def close(self):
+        """Close connection
+
+        Raises:
+            CommError: If any errors
+        """
+        pass
+
+    @abstractmethod
+    def send_frame(self, frame: BytesAlike):
+        """Send a SFM frame through the connection to the remote endpoint.
+
+        Args:
+            frame: The frame to be sent
+
+        Raises:
+            CommError: If any error happens while sending the frame
+        """
+        pass
+
+    def register_frame_receiver(self, receiver: FrameReceiver):
+        """Register frame receiver
+
+        Args:
+            receiver: The frame receiver
+
+        Raises:
+            CommError: If any error happens while processing the frame
+        """
+        self.frame_receiver = receiver
+
+    def process_frame(self, frame: BytesAlike):
+        """A convenience function to call frame receiver
+
+        Args:
+            frame: The frame to be processed
+
+        Raises:
+            CommError: If any error happens while processing the frame
+        """
+
+        if self.frame_receiver:
+            self.frame_receiver.process_frame(frame)
+        else:
+            log.error(f"Frame receiver not registered for {self}")
+
+    def __str__(self):
+
+        if self.state != ConnState.CONNECTED:
+            return f"[{self.name} Not Connected]"
+
+        conn_props = self.get_conn_properties()
+        local_addr = conn_props.get(DriverParams.LOCAL_ADDR, "N/A")
+        peer_addr = conn_props.get(DriverParams.PEER_ADDR, "N/A")
+        direction = "=>" if self.connector.mode == Mode.ACTIVE else "<="
+        peer_cn = conn_props.get(DriverParams.PEER_CN, None)
+        cn = " SSL " + peer_cn if peer_cn else ""
+        return f"[{self.name} {local_addr} {direction} {peer_addr}{cn}]"
```

Comment:
How does a `Connection` class different from `Endpoint` class?

Why both of them having these connections properties?

Maybe help me with an example,
if we have two parties: party A and party B
If party A is listening, party B is requesting.
We will have 2 Connections objects.
One at party A and its connection properties has ``local_addr: {A's ip}, peer_addr: {B's ip}``
One at party B and its connection properties has ``local_addr: {B's ip}, peer_addr: {A's ip}``
And we will have 2 Endpoint objects
One at party A and its connection properties has ``local_addr: {A's ip}, peer_addr: {B's ip}``
One at party B and its connection properties has ``local_addr: {B's ip}, peer_addr: {A's ip}``

Am I understanding it correctly?

Reply:
An endpoint can have multiple connections. 
conn_props in the endpoint is just a convenience. It's a superset of the properties of all its connections.

There is a tricky issue here. Say we have 2 connections for the same endpoint, one is over SSL and the other is not. The endpoint is treated as secure now since certificate is available from one of the connections.

Reply:
> An endpoint can have multiple connections.

Got it, this makes sense to me.

> conn_props in the endpoint is just a convenience. It's a superset of the properties of all its connections.
There is a tricky issue here. Say we have 2 connections for the same endpoint, one is over SSL and the other is not. The endpoint is treated as secure now since certificate is available from one of the connections.

In this case, should we just not have `conn_props` in the endpoint, that is really confusing ^

Reply:
Let's make this change later. In the old design, the message callback only gets endpoint, not connections. The connection is added to the callback now so we can do this. But it requires some Cellnet changes. Let's do it when Cellnet is merged.

Reply:
Got it, sounds good.